dlib/matrix/matrix_fft_abstract.h

```diff
@@ -231,6 +231,155 @@ namespace dlib
                   inverse transformation.  
     !*/
 
+// ----------------------------------------------------------------------------------------
+
+    template <typename EXP>
+    auto stft (
```

Comment:
```suggestion
    typename EXP::matrix_type stft (
```
Or something if this isn't right to say what the type is in some way.

Reply:
Although I guess this is possibly confusing since the other template arguments of `matrix` might be set differently.  Maybe the docs you have are fine.  At least make this return like `matrix` or something though.   This file doesn't have to compile, just be maximally informative to the reader.

Reply:
Yeah no worries. Wouldn't it be better if we just added the docs to the main header and forget about the abstract files? When Ctrl+clicking through an IDE, you never see the abstract files (at least I don't), so having the docs near the code is actually more useful. But we should be consistent with the rest of the library...

Reply:
It depends.  If the main header is sufficiently trivial (i.e. there isn't that much code in it) then it's fine.  Since someone reading it is seeing mostly the docs.   But for headers that are full of code and other implementation details it's real distracting to readers.  They just scroll around and are like "what should I look at?  what are the public functions?  What functions are even provided?"

And yeah, we should be consistent with the rest of the library.  But really the main thing is documentation readability.  Like all the matrix expression code is real unreadable to people.  For example, the list of trig functions on matrices are in the docs like this
```
    const matrix_exp cos (
        const matrix_exp& m
    );
    /*!
        requires
            - matrix_exp::type == float, double, or long double 
        ensures
            - returns a matrix R such that:
                - R::type == the same type that was in m
                - R has the same dimensions as m
                - for all valid r and c:
                  R(r,c) == std::cos(m(r,c)) 
    !*/
```
rather than as the zany preprocessor macros and expression template classes that they actually are.

So for example, it's fine if there is no abstract header if the real header is sufficiently clean and easy to read as docs.  Like if the functions are just so short and simple they are not distracting.  Or if the code is organized with prototypes at the top of the header and then all the implementation details shoved down at the bottom past some barrier that clearly indicates that only implementation details users shouldn't look at are below it.  Although if you are going to do that kind of thing you might as well just put it in two files (the normal one and the abstract one) since then you don't have to bend over backwards trying to figure out how to split it up like that.  Which is sometimes easy and sometimes tricky.

You can't find the docs in your IDE?  IDK, you should configure your IDE slightly differently :) :shrug: